Sources/SwiftLexicalScopes/Scope.swift

```diff
@@ -0,0 +1,139 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import SwiftSyntax
+
+extension SyntaxProtocol {
+  /// Scope at the syntax node. Could be inherited from parent or introduced at the node.
+  var scope: Scope? {
+    switch self.syntaxNodeType {
+    case is SourceFileSyntax.Type:
+      FileScope(syntax: self)
+    default:
+      parent?.scope
+    }
+  }
+}
+
+protocol Scope {
+  var parent: Scope? { get }
+
+  var sourceSyntax: SyntaxProtocol { get }
+
+  func getDeclarationFor(name: String, at syntax: SyntaxProtocol) -> Syntax?
```

Comment:
Perhaps we should call this `lookup` and have it return an array of `DeclSyntax` nodes. It is possible to shadow declarations across Scopes in Swift, and I would expect a lookup in a nested scope to be able to see both declarations. That makes it the client's responsibility to overlay a tie-breaking policy (e.g. deepest decl wins)

Reply:
Oh yeah, I haven’t thought about this case. You’re right, thanks for pointing this out. I adjusted the `Scope` protocol accordingly.